File: python/pyjse/ast/parser.py
"""JSON to AST parser for JSE.

Following the design in docs/regular.md:
- Recognizes symbols (strings starting with $ but not $$)
- Identifies expressions in arrays and objects
- Creates AST nodes with proper environment capture
"""

from pyjse.env import Env
from pyjse.types import JseValue
from pyjse.ast.nodes import (
    SymbolNode,
    ArrayNode,
    ObjectNode,
    ExpressionNode,
    QuoteNode,
    LiteralNode,
)


def _is_symbol(s: str) -> bool:
    """Check if string is a JSE symbol.

    Symbols start with $ but not $$.
    Wildcard $* is not a symbol.

    Args:
        s: String to check

    Returns:
        True if string is a symbol
    """
    # Wildcard $* is a literal string, not a symbol
    if s == "$*":
        return False
    return s.startswith("$") and not s.startswith("$$")


def _unescape(s: str) -> str:
    """Unescape $$-prefixed string.

    Args:
        s: String to unescape

    Returns:
        Unescaped string
    """
    if s.startswith("$$"):
        return s[1:]
    return s


class Parser:
    """Parse JSON values into JSE AST nodes.

    The parser identifies expressions based on the formal semantics:
    - ArrayExpr: [symbol, args...] where first element is a symbol
    - ObjectExpr: {symbol: value} where exactly one key is a symbol
    - QuoteExpr: ["$quote", x] or {"$quote": x} (special case)
    """

    def __init__(self, env: Env) -> None:
        """Initialize parser with environment.

        Args:
            env: Environment to pass to constructed AST nodes
        """
        self._env = env

    def parse(self, expr: JseValue) -> JseValue:
        """Parse JSON value into AST node or return primitive.

        Args:
            expr: JSON value to parse

        Returns:
            AST node or primitive value
        """
        # Primitives - return as-is wrapped in LiteralNode
        if expr is None or isinstance(expr, (int, float, bool)):
            return LiteralNode(expr, self._env)

        # Strings - check if symbol
        if isinstance(expr, str):
            if _is_symbol(expr):
                return SymbolNode(expr, self._env)
            return LiteralNode(_unescape(expr), self._env)

        # Arrays - check for expression form
        if isinstance(expr, list) or isinstance(expr, tuple):
            return self._parse_list(expr)

        # Objects - check for expression form
        if isinstance(expr, dict):
            return self._parse_dict(expr)

        # Unknown type - wrap as literal
        return LiteralNode(expr, self._env)

    def _parse_list(self, lst: list) -> JseValue:
        """Parse list expression.

        Forms:
        - ["$quote", x] -> QuoteNode
        - [symbol, ...] -> ArrayNode (function call)
        - [...] -> ArrayNode (regular array)

        Args:
            lst: List to parse

        Returns:
            AST node
        """
        if not lst:
            # Empty list
            return ArrayNode([], self._env)

        first = lst[0]

        # Special case: $quote
        if first == "$quote":
            # ["$quote", x] - return x unevaluated
            value = lst[1] if len(lst) > 1 else None
            return QuoteNode(value, self._env)

        # Check if this is a function call form
        if isinstance(first, str) and _is_symbol(first):
            # [symbol, args...] - function call
            elements = [self.parse(e) for e in lst[1:]]
            return ExpressionNode(first, elements, {}, self._env)

        # Regular array - evaluate all elements
        elements = [self.parse(e) for e in lst]
        return ArrayNode(elements, self._env)

    def _parse_dict(self, d: dict) -> JseValue:
        """Parse dict expression.

        Forms:
        - {"$quote": x} -> QuoteNode
        - {symbol: value, ...} -> ObjectNode (exactly one symbol key)
        - {...} -> ObjectNode with regular keys

        Args:
            d: Dict to parse

        Returns:
            AST node or dict of parsed values
        """
        # Find symbol keys
        symbol_keys = [k for k in d if _is_symbol(k)]

        if len(symbol_keys) == 0:
            # Regular object - parse all values
            # Note: This creates a dict, not an ObjectNode
            result = {}
            for k, v in d.items():
                parsed_key = _unescape(k) if isinstance(k, str) else k
                parsed_value = self.parse(v)
                # Store the actual parsed value (might be AST node)
                result[parsed_key] = parsed_value
            return ObjectNode(result, self._env)

        if len(symbol_keys) == 1:
            # Operator form: {"$operator": value, "meta": ...}
            operator = symbol_keys[0]

            # Special case: $quote
            if operator == "$quote":
                value = d[operator]
                return QuoteNode(value, self._env)

            # Parse the value
            parsed_value = self.parse(d[operator])

            # Parse metadata (other keys)
            metadata = {}
            for k, v in d.items():
                if k != operator:
                    parsed_key = _unescape(k) if isinstance(k, str) else k
                    parsed_metadata = self.parse(v)
                    metadata[parsed_key] = parsed_metadata

            return ExpressionNode(operator, parsed_value, metadata, self._env)

        # Multiple symbol keys - error per formal semantics
        raise ValueError("JSE structure error: object cannot have multiple operator keys")

File: java/src/main/java/io/github/marchliu/jse/Parser.java
package io.github.marchliu.jse;

import io.github.marchliu.jse.ast.*;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Parser for converting JSON values to AST nodes.
 *
 * <p>Follows the formal semantics:
 * <ul>
 *   <li>Symbols: strings starting with {@code $} but not {@code $$} (except {@code $*} wildcard)</li>
 *   <li>Arrays: parsed as ArrayNode, with special handling for {@code $quote}</li>
 *   <li>Objects: parsed as ObjectNode or ObjectExpressionNode</li>
 * </ul>
 */
public class Parser {

    private final Env env;

    /**
     * Create a parser with the given environment.
     *
     * @param env Environment to pass to constructed AST nodes
     */
    public Parser(Env env) {
        this.env = env;
    }

    /**
     * Parse a JSON value into an AST node or return primitive.
     *
     * @param expr JSON value to parse
     * @return AST node or primitive value
     */
    public Object parse(Object expr) {
        // Primitives - wrap in LiteralNode
        if (expr == null || expr instanceof Number || expr instanceof Boolean) {
            return new LiteralNode(expr, env);
        }

        // Strings - check if symbol
        if (expr instanceof String s) {
            if (isSymbol(s)) {
                return new SymbolNode(s, env);
            }
            return new LiteralNode(unescape(s), env);
        }

        // Lists - check for expression form
        if (expr instanceof List<?> list) {
            return parseList(list);
        }

        // Objects - check for expression form
        if (expr instanceof Map<?, ?> map) {
            @SuppressWarnings("unchecked")
            Map<String, Object> obj = (Map<String, Object>) map;
            return parseDict(obj);
        }

        // Unknown type - wrap as literal
        return new LiteralNode(expr, env);
    }

    /**
     * Parse list expression.
     *
     * <p>Forms:
     * <ul>
     *   <li>{@code ["$quote", x]} -> QuoteNode</li>
     *   <li>{@code [symbol, ...]} -> ArrayNode (function call)</li>
     *   <li>{@code [...]} -> ArrayNode (regular array)</li>
     * </ul>
     */
    private Object parseList(List<?> list) {
        if (list.isEmpty()) {
            return new ArrayNode(new ArrayList<>(), env);
        }

        Object first = list.get(0);

        // Special case: $quote
        if ("$quote".equals(first)) {
            Object value = list.size() > 1 ? list.get(1) : null;
            return new QuoteNode(value, env);
        }

        // Check if this is a function call form
        if (first instanceof String s && isSymbol(s)) {
            List<Object> elements = new ArrayList<>(list.size());
            for (Object item : list) {
                elements.add(parse(item));
            }
            return new ArrayNode(elements, env);
        }

        // Regular array - parse all elements
        List<Object> elements = new ArrayList<>(list.size());
        for (Object item : list) {
            elements.add(parse(item));
        }
        return new ArrayNode(elements, env);
    }

    /**
     * Parse dict expression.
     *
     * <p>Forms:
     * <ul>
     *   <li>{@code {"$quote": x}} -> QuoteNode</li>
     *   <li>{@code {symbol: value}} -> ObjectExpressionNode (exactly one symbol key)</li>
     *   <li>{@code {...}} -> ObjectNode with regular keys</li>
     * </ul>
     */
    private Object parseDict(Map<String, Object> dict) {
        // Find symbol keys
        List<String> symbolKeys = new ArrayList<>();
        for (String key : dict.keySet()) {
            if (isSymbol(key)) {
                symbolKeys.add(key);
            }
        }

        if (symbolKeys.isEmpty()) {
            // Regular object - parse all values
            Map<String, Object> result = new LinkedHashMap<>();
            for (Map.Entry<String, Object> entry : dict.entrySet()) {
                String unescapedKey = unescape(entry.getKey());
                result.put(unescapedKey, parse(entry.getValue()));
            }
            return new ObjectNode(result, env);
        }

        if (symbolKeys.size() == 1) {
            // Operator form: {"$operator": value, "meta": ...}
            String operator = symbolKeys.get(0);

            // Special case: $quote
            if ("$quote".equals(operator)) {
                return new QuoteNode(dict.get(operator), env);
            }

            // Parse the value
            Object parsedValue = parse(dict.get(operator));

            // Parse metadata (other keys)
            Map<String, Object> metadata = new LinkedHashMap<>();
            for (Map.Entry<String, Object> entry : dict.entrySet()) {
                if (!entry.getKey().equals(operator)) {
                    String unescapedKey = unescape(entry.getKey());
                    metadata.put(unescapedKey, parse(entry.getValue()));
                }
            }

            return new ObjectExpressionNode(operator, parsedValue, metadata, env);
        }

        // Multiple symbol keys - error per formal semantics
        throw new IllegalArgumentException("JSE structure error: object cannot have multiple operator keys");
    }

    /**
     * Check if string is a JSE symbol.
     *
     * <p>Symbols start with {@code $} but not {@code $$}.
     * Wildcard {@code $*} is NOT a symbol.</p>
     *
     * @param s String to check
     * @return true if string is a symbol
     */
    public static boolean isSymbol(String s) {
        // Wildcard $* is a literal string, not a symbol
        if ("$*".equals(s)) {
            return false;
        }
        return s.startsWith("$") && !s.startsWith("$$");
    }

    /**
     * Unescape {@code $$}-prefixed string.
     *
     * @param s String to unescape
     * @return Unescaped string
     */
    private static String unescape(String s) {
        if (s.startsWith("$$")) {
            return s.substring(1);
        }
        return s;
    }
}
